### PR TITLE
Fix #3170: Get old value when update on rows with big row_id

### LINKED ISSUE #3170: update error
#### What happens?
when id bigger than 122880 update successful but get older value.

#### To Reproduce
`@Test 
   public void duckDBUpdateFailTest() throws SQLException { 

       try (DuckDBConnection conn = (DuckDBConnection) DriverManager.getConnection("jdbc:duckdb:")) { 
           Statement stmt = conn.createStatement(); 
           stmt.execute("CREATE SCHEMA IF NOT EXISTS mydb"); 
           stmt.execute("CREATE TABLE IF NOT EXISTS mydb.student(id INTEGER,name VARCHAR,PRIMARY KEY(id))"); 
           DuckDBAppender appender = conn.createAppender("mydb", "student"); 
           for (int i = 0; i < 130001; i++) { 
               appender.beginRow(); 
               appender.append(i); 
               appender.append("creator"); 
               appender.endRow(); 
               if (i % 10000 == 0) { 
                   appender.flush(); 
               } 
           } 
           appender.close(); 

           int id = 122881; 
           System.out.println("=================id=" + id + "==================="); 
           String selectSql = "select * from mydb.student where id = %d"; 
           ResultSet rs = stmt.executeQuery(String.format(selectSql, id)); 
           System.out.println("=================before update==================="); 
           while (rs.next()) { 
               for (int i = 1; i <= rs.getMetaData().getColumnCount(); i++) { 
                   System.out.print(rs.getObject(i)); 
                   System.out.print("\t"); 
               } 
           } 

           //update 
           System.out.println("\n"); 
           System.out.println("update mydb.student set name = updator where id = " + id); 
           String upadteSql = "update mydb.student set name = ? where id = ?"; 
           PreparedStatement updateStmt = conn.prepareStatement(upadteSql); 
           updateStmt.setObject(1, "updator"); 
           updateStmt.setObject(2, id); 
           int updateResult = updateStmt.executeUpdate(); 
           System.out.println("\n"); 
           System.out.println("==============update result:" + updateResult); 
           //update 
           rs = stmt.executeQuery(String.format(selectSql, id)); 
           System.out.println("=================after update==================="); 
           while (rs.next()) { 
               for (int i = 1; i <= rs.getMetaData().getColumnCount(); i++) { 
                   System.out.print(rs.getObject(i)); 
                   System.out.print("\t"); 
               } 
           } 
       } 

   }
=================id=122881===================
=================before update=================== 
122881     creator      

update mydb.student set name = updator where id = 122881 

==============update result:1 
=================after update=================== 
122881     creator

=================id=122879===================
=================before update=================== 
122879     creator      

update mydb.student set name = updator where id = 122879 

==============update result:1 
=================after update=================== 
122879     updator      
      

`

#### Environment (please complete the following information):
 - OS: windows
 - DuckDB Version: 0.3.1
 - DuckDB Client: jdbc 0.3.1

### PR DESCRIPTION
**Pull request by zippond**
Symptom
===============
fix #3170. When id bigger than 122880 update successful but get older value.

Root Cause
===============
A segment/rowgroup stores 122880 rows. In UpdateSegment::FetchRow(), when the updated row is not in the first segment/rowgroup, the row_in_vector was incorrectly caculated which not take column_data.start into account.
In this case, the updated row cannot be fetched and old value was returned.

Fix
===============
row_in_vector calculation should be based on column_data.start.

Test 
===============
New tests added.

### CODE REVIEW

File: test/sql/update/test_update_issue_3170.test [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+UPDATE student SET name = 'updator0' WHERE id = 122879
+
+# rowgroup 1
+statement ok
+UPDATE student SET name = 'updator1' WHERE id = 122881
+
+# rowgroup 2 
+statement ok
+UPDATE student SET name = 'updator2' WHERE id = 245780
+
```
Comment (taniabogatsch) [inline test/sql/update/test_update_issue_3170.test:46]: Can we add another test here that reaches into a vector that does not have a `vector_index` of zero? Just for my sanity...
```sql
statement ok
UPDATE student SET name = 'updator3' WHERE id = 150881

query I
SELECT name FROM student WHERE id = 150881
----
updator3
```

File: test/sql/update/test_update_issue_3170.test [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+query I
+SELECT name FROM student WHERE id = 122881
+----
+updator1
+
+# rowgroup 2
+query I
+SELECT name FROM student WHERE id = 245780
+----
+updator2
```
Comment (taniabogatsch) [inline test/sql/update/test_update_issue_3170.test:64]: Can we also add the original reproducible example as another test?

```sql
load __TEST_DIR__/list_index_compression.db

statement ok
CREATE TABLE student(id INTEGER,name VARCHAR,PRIMARY KEY(id));

statement ok
insert into student select i, 'creator' from range(130001) tbl(i);

query II
select id, name from student where id=122881;
----
122881	creator

statement ok
update student set name = 'updator' where id = 122881;

query II
select id, name from student where id=122881;
----
122881	updator

restart

query II
select id, name from student where id=122881;
----
122881	updator
```

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Thanks, @zippond, for spotting this and for opening the PR. I've looked into it, and I'll leave my comments here for the person potentially merging this.

As I've not worked much with storage before, here are my general assumptions. I am honestly surprised that this has not yet caused many more bugs. (But maybe I am missing something, therefore, my extensive comment.)
- `column_data.start` is always a multiple of our row group size, i.e., 122880.
- The `row_id` exceeds the row group size and can be any
- We store the data in consecutive vectors of `STANDARD_VECTOR_SIZE`

There are two crucial calculations here (1 and 2/3):
```c++
// 1.
idx_t vector_index = (row_id - column_data.start) / STANDARD_VECTOR_SIZE;
// [...]
// 2. old
idx_t row_in_vector = row_id - vector_index * STANDARD_VECTOR_SIZE;
// 3. new
idx_t row_in_vector = (row_id - column_data.start) - vector_index * STANDARD_VECTOR_SIZE;
```

1.: We subtract `column_data.start` from the `row_id`, so we look into the current row group. Then, we calculate the `vector_index` within that row group.
2.: We now try to get the row in our row group. But since we do not subtract the row group offset `column_data.start` first, we calculate the `row_in_vector` position in row group 0, not the current one.
3.: Fixes this miscalculation by subtracting the offset of the current row group from the `row_id`.

I also checked out the branch and am running the nightly tests on my fork. After adding my two suggested tests, this can be merged if the nightly runs pass.

File: test/sql/update/test_update_issue_3170.test (hunk 1 above)
Comment (zippond) [reply test/sql/update/test_update_issue_3170.test:52]: Thank you Tania, I just added this test.

File: test/sql/update/test_update_issue_3170.test (hunk 2 above)
Comment (zippond) [reply test/sql/update/test_update_issue_3170.test:75]: I didn't use the "load/restart" before, so this test will start a on-file db and test with a restart, right?
I just appended the test to the end, let me know if I misunderstood anything.

**Review by taniabogatsch — APPROVED** [review]
Looks good to me now, thanks!
The CI (and the nightly runs) are also passing on this ([here](https://github.com/taniabogatsch/duckdb/tree/nightly-run-pr-8520)).

**Pull request merged**